### PR TITLE
prepare upstream for downstream (WIP, for #341)

### PR DESCRIPTION
**Pull request by leo19941227**
Hi @daisukelab and @andi611 ,

Let's move the discussion here! (from #341)
Sorry I was busy yesterday and can't find time for this adjustment.
After digging more into @daisukelab's needs, I think the problem is not about how to pass downstream config into upstream model, since even this is made available, the upstream implementation would need to take care of setting up the downstream (like parsing the downstream dataset for every task) which seems not realistic too me, and can be inconsistent with the official downstream implementation, violating the SSOT principle.

Hence, I think it might be more appropriate to pass the already parse & prepared downstream dataloaders, including train/valid/test dataloaders prepared by `DownstreamExpert` via the `get_dataloader` interface.
By simply iterating over the dataloader, the upstream model can get all the batches of waveforms. The first argument of each batch is a list of waveforms, in the format of `List[Union[torch.FloatTensor, np.ndarray]]`. Then, the upstream is now possible to pre-extract all the waveforms and calculate the statistics.

Does @daisukelab think this will be a feasible solution?
Please let me know if I miss anything. Thanks!

Leo

### CODE REVIEW

**Comment by daisukelab** [comment]
@leo19941227 Hi, thank you for handling this issue!
It is a lovely solution. (BTW, it is also what I have been doing with my linear evaluation.)
I can also suggest that a dataloader for the training set would be more appropriate than that for the test set, whether in the training or test phase. It is for limiting the model to access only the training data distribution in advance.
But if preparing the training dataloader is difficult, I think a test dataloader would be acceptable since it can also be a possible use case.

File: s3prl/upstream/wav2vec2/expert.py [hunk 1]
```diff
@@ -56,6 +56,9 @@ def postprocess(xs):
     def get_downsample_rates(self, key: str) -> int:
         return 320
 
+    def prepare_for_downstream(self, split: str, dataloader):
```
[the remark this answers is not in the document]
Comment (leo19941227) [reply s3prl/upstream/wav2vec2/expert.py:59]: Oh it comes to me that, in the current API, the upstream model does not know which split's statistics it should use when doing forward... I think I would follow your suggestion to always pass the train dataloader into this `prepare_for_downstream`. It is also easier to understand for the user.

**Comment by leo19941227** [comment]
Hi @daisukelab !

I adjust the API a little bit, so that the upstream now takes a list of waveforms from the train dataloader, and can configure the stats it needs. Please have a look and let me know whether this fits your need!

P.S. I think passing a list of waveform can make this upstream more easily used in other scenario for the non-SUPERB users, or else they will need to prepare a dataloader in the SUPERB's format just to configure the upstream's stats.

Leo

**Comment by leo19941227** [comment]
If you are good with this new API, I will merge this for you to move on to the official BYOL-A implementation :)

**Comment by daisukelab** [comment]
@leo19941227  Thank you for your quick work!
I just finished a simple test with BYOL-A changes at my end, so I will incorporate this API implementation and try it out.
I can probably do it today. I will get back to you!

**Comment by leo19941227** [comment]
Cool! Sounds good to me!

**Comment by leo19941227** [comment]
Hi @daisukelab !

Just a minor follow-up question, are you still planning to make this pull request?
Or perhaps we can finish this part ourselves first and then pin you for a review?
I am good with either ways.

Thanks!

**Comment by daisukelab** [comment]
@leo19941227 Hi, I was waiting for you actually :)
Can you find my comment regarding the error above? It stops us from moving forward.

**Comment by leo19941227** [comment]
Hey @daisukelab ,

Oh I did not know that. Would you mind let me know again what is the error?

**Comment by daisukelab** [comment]
@leo19941227 Excuse me for the belated reply. (being busy for DCASE challenge ops these days)
- An error happens at line 280 in downstream/runner.py if I add the `configure_stats` method in byola/expert.py.
    def configure_stats(self, wavs: List[Union[torch.Tensor, np.ndarray]]):
- Which can be fixed if we remove `, epoch=0` from line 280.

**Comment by leo19941227** [comment]
No problem! Not in hurry. I will fix this, thanks for the information.

**Comment by daisukelab** [comment]
@leo19941227 Dear Leo, shall we restart solving this ...? I understand that the current code is fairly different from what we are working on this. It could be better be starting over, ... just my guess.

**Comment by leo19941227** [comment]
Hi @daisukelab ,

Sure! You are right, and sorry for my delay. I was busy with other activities, while I am available now for this issue.
Let's fix this. I would adjust the code according to your suggestion. Thanks for the reminder!

Sincerely,
Leo

**Comment by daisukelab** [comment]
@leo19941227 Hi Leo, it's good news. We improved version BYOL-A implementation in the pull request: https://github.com/s3prl/s3prl/pull/455
It doesn't require extending the SUPERB system to let the upstream models pre-compute statistics. Ours do it online.
We hope you like it. Please let me know of any issues.
Thank you.
